pgmpy/inference/continuous/base.py

```diff
@@ -0,0 +1,365 @@
+
+
+class BaseSimulateHamiltonianDynamics(object):
+    """
+    Base class for proposing new values of position and momentum by simulating Hamiltonian Dynamics.
+
+    Classes inheriting this base class can be passed as an argument for
+    simulate_dynamics in inference algorithms.
+
+    Parameters
+    ----------
+    grad_log_pdf : A subclass of pgmpy.inference.continuous.BaseGradLogPDF
+
+    model : An instance of pgmpy.models.Continuous
+        Model for which DiscretizeTime object is initialized
+
+    position : A 1d array like object
+        Vector representing values of parameter position( or X)
+
+    momentum: A 1d array like object
+        Vector representing the proposed value for momentum (velocity)
+
+    stepsize: Float
+        stepsize for the simulating dynamics
+
+    grad_log_position: A 1d array like object, defaults to None
+        Vector representing gradient log at given position
+        If None, then will be calculated
+
+    Examples
+    --------
+    >>> from pgmpy.inference.continuous import BaseSimulateHamiltonianDynamics
+    >>> from pgmpy.models import JointGaussianDistribution
+    >>> from pgmpy.inference.continuous import GradLogPDFGaussian
+    >>> # Class should initalize self.new_position, self.new_momentum and self.new_grad_logp
+    >>> # self.new_grad_logp represents gradient log at new proposed value of position
+    >>> class ModifiedEuler(BaseSimulateHamiltonianDynamics):
+    ...     def __init__(self, grad_log_pdf, model, position, momentum, stepsize, grad_log_position=None):
+    ...         BaseSimulateHamiltonianDynamics.__init__(self, grad_log_pdf, model, position,
+    ...                                                  momentum, stepsize, grad_log_position)
+    ...         self.new_position, self.new_momentum, self.new_grad_logp = self._get_proposed_values()
+    ...     def _get_proposed_values(self):
+    ...         momentum_bar = self.momentum + self.stepsize * self.grad_log_position
+    ...         position_bar = self.position + self.stepsize * momentum_bar
+    ...         return position_bar, momentum_bar, grad_log_position
+    >>> import numpy as np
+    >>> pos = np.array([[1], [2]])
+    >>> momentum = np.array([[0], [0]])
+    >>> mean = np.array([[0], [0]])
+    >>> covariance = np.eye(2)
+    >>> model = JointGaussianDistribution(mean, covariance)
+    >>> new_pos, new_momentum, new_grad = ModifiedEuler(GradLogPDFGaussian, model,
+    ...                                                 pos, momentum, 0.25).get_proposed_values()
+    >>> new_pos
+    array([[ 0.9375],
+           [ 1.875 ]])
+    >>> new_momentum
+    array([[-0.25],
+           [-0.5 ]])
+    >>> new_grad
+    array([[-1.],
+           [-2.]])
+    """
+
+    def __init__(self, grad_log_pdf, model, position, momentum, stepsize, grad_log_position=None):
+
+        if isinstance(position, (np.matrix, np.ndarray, list, tuple, set, frozenset)):
+            position = np.array(position).flatten()
```

Comment:
@ankurankan My bad, I thought I'll be allowing more flexibility of passing any homogenous array type, and will then simply reshape it, in the way I need. I'll change it. This thing will create ambiguity.